packages/skeleton/src/components/project.ts

```diff
@@ -0,0 +1,462 @@
+import fs from 'fs'
+import path from 'path'
+
+import chalk from 'chalk'
+
+import { getPaths } from '@redwoodjs/internal/dist/index'
+
+import { getRootPath } from '../lib/path'
+
+import { extractCells } from './cell'
+import type { RedwoodCell } from './cell'
+import { RedwoodError, RedwoodWarning } from './diagnostic'
+import { extractDirectives } from './directive'
+import type { RedwoodDirective } from './directive'
+import { extractFunctions } from './function'
+import type { RedwoodFunction } from './function'
+import { extractLayouts } from './layout'
+import type { RedwoodLayout } from './layout'
+import { extractPages } from './page'
+import type { RedwoodPage } from './page'
+import { extractRouters } from './router'
+import type { RedwoodRouter } from './router'
+import { extractSDLs } from './sdl/sdl'
+import type { RedwoodSDL } from './sdl/sdl'
+import { extractServices } from './service/service'
+import type { RedwoodService } from './service/service'
+import { extractSides } from './side'
+import type { RedwoodSide } from './side'
+import { RedwoodSkeleton } from './skeleton'
+import { extractTOMLs } from './toml'
+import type { RedwoodTOML } from './toml'
+
+/**
+ * Used to enumerate either JS or TS project types
+ */
+export enum RedwoodProjectType {
+  TYPESCRIPT = 'typescript',
+  JAVASCRIPT = 'javascript',
+}
+
+export class RedwoodProject extends RedwoodSkeleton {
+  warnings: RedwoodWarning[] = []
+  errors: RedwoodError[] = []
+
+  readonly type: RedwoodProjectType
+
+  private cells?: RedwoodCell[] | undefined
+  private directives?: RedwoodDirective[] | undefined
+  private functions?: RedwoodFunction[] | undefined
+  private layouts?: RedwoodLayout[] | undefined
+  private pages?: RedwoodPage[] | undefined
+  private routers?: RedwoodRouter[] | undefined
+  private sdls?: RedwoodSDL[] | undefined
+  private services?: RedwoodService[] | undefined
+  private sides?: RedwoodSide[] | undefined
+  private tomls?: RedwoodTOML[] | undefined
+
+  public static getProject({
+    pathWithinProject = '',
+    full = false,
+    readFromCache = true,
+    insertIntoCache = true,
+  }: {
+    pathWithinProject?: string
+    full?: boolean
+    readFromCache?: boolean
+    insertIntoCache?: boolean
+  } = {}) {
+    const projectCache = RedwoodProjectsCache.getInstance().projects
+    const rootPath = getRootPath(pathWithinProject)
+    if (readFromCache) {
+      const cachedProject = projectCache.get(rootPath)
+      if (cachedProject) {
+        return cachedProject
+      }
+    }
+    const project = new RedwoodProject(rootPath, full)
+    if (insertIntoCache) {
+      projectCache.set(project.filepath, project)
+    }
+    return project
+  }
+
+  private constructor(rootPath: string, full = false) {
```

Comment:
Doesn't seem necessary? I'm not quite sure yet. What do you think?

```suggestion
  constructor(rootPath: string, full = false) {
```

Reply:
I didn't want to let instances of project be created without interacting with the caching mechanism so I made the constructor private.